### DISCUSSION TITLE
Type hinting ExceptionGroup pytest.raises function

### DISCUSSION THREAD

**Discussion by machadinhos** [microsoft/pyright#8270] [Q&A] [answered]
Hi,

While writing some tests, I encountered an issue with Pyright, which is reporting partially unknown types, and I'm not sure how to fix it.

Here is my test function:

```py
    with pytest.raises(ExceptionGroup) as exc_info:
        tablerone_partial(
            padding_horizontal_no_border=paddings[0],
            padding_vertical_no_border=paddings[1],
            padding_horizontal_with_border=paddings[2],
            padding_vertical_with_border=paddings[3],
        )
    assert all(isinstance(exc, InvalidPaddingValueError) for exc in exc_info.value.exceptions)

```
In this case, I'm trying to ensure that tablerone_partial raises an ExceptionGroup that only contains `InvalidPaddingValueError`. The `ExceptionGroup` holds a more generic exception type, and `InvalidPaddingValueError` inherits from that generic exception type.

Here are the errors shown by Pyright:
```
 test_args.py:27:43 - error: Type of "exc_info" is partially unknown
    Type of "exc_info" is "ExceptionInfo[ExceptionGroup[Unknown]]" (reportUnknownVariableType)
  test_args.py:34:62 - error: Type of "exc" is partially unknown
    Type of "exc" is "Unknown | ExceptionGroup[Unknown]" (reportUnknownVariableType)
  test_args.py:34:69 - error: Type of "value" is partially unknown
    Type of "value" is "ExceptionGroup[Unknown]" (reportUnknownMemberType)
  test_args.py:34:69 - error: Type of "exceptions" is partially unknown
    Type of "exceptions" is "tuple[Unknown | ExceptionGroup[Unknown], ...]" (reportUnknownMemberType)
```
I've tried adding type hints to the `ExceptionGroup` raised by the `tablerone_partial` function, but it didn't help. (Mypy doesn't show any errors.)

How can I hint the `exc_info` to make Pyright recognize the type? Or should I be approaching this differently?

Thanks in advance for any help!

**Comment by erictraut** [comment]
Could you provide a self-contained minimal example? The code above refers to symbols that are not defined.

**Reply by machadinhos** [reply]
Sure. After some more testing I was able to isolate the error to this:
```py
def a() -> None:
    ex: ExceptionGroup[ValueError] = ExceptionGroup("", [ValueError("a"), ValueError("b")])
    raise ex

try:
    a()
except ExceptionGroup as e:
    print(all(isinstance(exc, ValueError) for exc in e.exceptions))
```
The `ex: ExceptionGroup[ValueError]` is just to give maximum context.

Pyright shows this errors:
```
  test.py:8:26 - error: Type of "e" is partially unknown
    Type of "e" is "ExceptionGroup[Unknown]" (reportUnknownVariableType)
  test.py:9:47 - error: Type of "exc" is partially unknown
    Type of "exc" is "Unknown | ExceptionGroup[Unknown]" (reportUnknownVariableType)
  test.py:9:54 - error: Type of "exceptions" is partially unknown
    Type of "exceptions" is "tuple[Unknown | ExceptionGroup[Unknown], ...]" (reportUnknownMemberType)
```

I believe the second and third error are derived by te first one because doing this at the try except block:
```py
except ExceptionGroup as e:
    pass
```
leads to just te first error:
```
  test.py:8:26 - error: Type of "e" is partially unknown
    Type of "e" is "ExceptionGroup[Unknown]" (reportUnknownVariableType)
```

**Reply by erictraut** [reply, marked as answer]
`ExceptionGroup` is a generic class that accepts one type argument. In this case, there's no way for pyright to determine the type argument, so it becomes `Unknown`. That is, the type of `e` is `ExceptionGroup[Unknown]`.

With [PEP ](https://peps.python.org/pep-0696/)696 now officially accepted, the maintainers of typeshed could specify that the type parameter for `ExceptionGroup` defaults to `Exception`. In that case, pyright will no longer assume `Unknown` in this case. If that sounds like a good solution, please file a request (or submit a PR) in the typeshed project.

In the meantime, you can avoid the error by specifying a type declaration for `e`.

```python
e: BaseExceptionGroup[Exception]
```

**Reply by machadinhos** [reply]
Thanks!

I'm not sure how to file a request (let alone submit a PR) to the typeshed project...I don't understand how static type checkers work...
Any idea where I can get help to file that request?

PEP 696 will only take effect in python 3.13, should this be already brought up?

**Reply by erictraut** [reply]
The [typeshed project](https://github.com/python/typeshed) provides type stubs that define the types for stdlib symbols like [ExceptionGroup](https://github.com/python/typeshed/blob/dcab6e88883c629ede9637fb011958f8b4918f52/stdlib/builtins.pyi#L2053).

The change would involve a change to [this line](https://github.com/python/typeshed/blob/dcab6e88883c629ede9637fb011958f8b4918f52/stdlib/builtins.pyi#L2011).

PEP 696 backward compatibility is provided by typing_extensions at runtime. Type stubs don't require runtime support, so that's not an issue here. The typeshed stubs have already started to incorporate support for it (e.g. [here](https://github.com/python/typeshed/blob/dcab6e88883c629ede9637fb011958f8b4918f52/stdlib/typing.pyi#L450)).

**Reply by machadinhos** [reply]
I've opened an [issue](https://github.com/python/typeshed/issues/12245) in the typeshed repository.

Thanks a lot for all the help!
